dbms/src/DataStreams/LimitTransformAction.cpp

```diff
@@ -17,51 +17,53 @@
 
 namespace DB
 {
-LimitTransformAction::LimitTransformAction(
-    const Block & header_,
-    size_t limit_)
-    : header(header_)
-    , limit(limit_)
+namespace
 {
-}
-
-Block LimitTransformAction::getHeader() const
+void cut(Block & block, size_t rows, size_t limit, size_t pos)
 {
-    return header;
+    // give away a piece of the block
+    assert(rows + limit > pos);
+    size_t length = rows + limit - pos;
+    for (size_t i = 0; i < block.columns(); ++i)
+        block.safeGetByPosition(i).column = block.safeGetByPosition(i).column->cut(0, length);
```

Comment:
```cpp
for (auto & col : block)
{
   auto mutate_col = (*std::move(col.column)).mutate();
   mutate_col->popBack(pos - limit);
   col.column = std::move(mutate_col);
}
```

which can avoid the overhead of memory copy.

Reply:
updated.